src/util.ts

```diff
@@ -21,19 +21,34 @@ export function isMdEditor(editor: TextEditor) {
 }
 
 /**
- * Match some kinds of fenced code blocks.
+ * **Do not call `exec()` method, to avoid accidentally changing its state!**
+ *
+ * Match most kinds of fenced code blocks:
+ *
+ * * Only misses <https://spec.commonmark.org/0.29/#example-116>.
+ * * Due to the limitations of regular expression, the "end of the document" cases are not handled.
  */
-export const REGEX_FENCED_CODE_BLOCK = /^(?: {0,3}|\t)(?<fence>(?<char>[`~])\k<char>{2,})[^`\r\n]*$[^]*?^(?: {0,3}|\t)\k<fence>\k<char>* *$/gm;
+export const REGEX_FENCED_CODE_BLOCK = /^ {0,3}(?<fence>(?<char>[`~])\k<char>{2,})[^`\r\n]*$[^]*?^ {0,3}\k<fence>\k<char>* *$/gm;
 
 /**
- * Checks whether the line is in some kinds of fenced code blocks.
- * @param lineNum The zero-based line index.
+ * Checks whether the line is in a fenced code block.
+ * @param lineIndex The zero-based line index.
  */
-export function isInFencedCodeBlock(doc: TextDocument, lineNum: number): boolean {
-    let textBefore = doc.getText(new Range(new Position(0, 0), new Position(lineNum, 0)));
-    textBefore = textBefore.replace(REGEX_FENCED_CODE_BLOCK, '').replace(/<!--[^]+?-->/g, '');
-    //// So far `textBefore` should contain no valid fenced code block or comment
-    return /^( {0,3}|\t)```[^`\r\n]*$[^]*$/gm.test(textBefore);
+export function isInFencedCodeBlock(doc: TextDocument, lineIndex: number): boolean {
+    const docTokens = commonmarkEngine.engine.parse(doc.getText(), {});
+
+    for (const token of docTokens) {
+        if (
+            token.type === 'fence'
+            && token.tag === 'code'
+            && token.map![0] <= lineIndex
+            && lineIndex < token.map![1]
+        ) {
+            return true;
+        }
+    }
+
+    return false;
```

Comment:
This looks good.

Just curious what is the effect on the performance (e.g. documents larger than our default `sizeLimit`). And also any corner case that the previous tricky regex way cannot handle?

Reply:
> performance

* I did not experience notable lag when editing lists in a document with 58k characters (1k lines).
* markdown-it itself is already fast enough.

---

> case

A fenced code block can be deeply nested in other block structures. To recognize them correctly, **we have to parse the *whole* document**.

Example:

```markdown
> ~~~hello
> `world`
```

Reply:
👌

If there aren't obvious disadvantages, of course we should use this more principle way. Many thanks!